### PR TITLE
[SW-369] Use only Spark nodes with H2O for H2O-related computations

### PR DESCRIPTION
**Pull request by jakubhava**

### CODE REVIEW

**Comment by jakubhava** [comment]
Just reopened the old PR to see if it is still relevant

**Comment by jakubhava** [comment]
Attempting to write a test

File: core/src/main/scala/org/apache/spark/h2o/converters/WriteConverterCtxUtils.scala [hunk 1]
```diff
@@ -59,7 +62,7 @@ object WriteConverterCtxUtils {
     * Converts the RDD to H2O Frame using specified conversion function
     *
     * @param hc            H2O context
-    * @param rdd           rdd to convert
+    * @param rddInput           rdd to convert
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/converters/WriteConverterCtxUtils.scala]: nit: indentation

File: core/src/main/scala/org/apache/spark/rdd/h2o/H2OAwareRDD.scala [hunk 2]
```diff
@@ -14,20 +14,21 @@
...
-    Log.err(s"Executor ${executorAdded.executorId} without H2O instance discovered, killing the cloud!")
-    H2O.shutdown(-1)
-  }
+class H2OAwareRDD[U: ClassTag](nodes: Seq[String], prev: RDD[U]) extends RDD[U](prev: RDD[U]) {
+
+  override def getPreferredLocations(split: Partition): Seq[String] = nodes
 
+  override def compute(split: Partition, context: TaskContext): Iterator[U] = prev.compute(split, context)
+
+  override protected def getPartitions: Array[Partition] = {
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/rdd/h2o/H2OAwareRDD.scala]: nit: I would prefer one line without curly brackets

File: core/src/main/scala/org/apache/spark/rdd/h2o/H2OAwareRDD.scala [hunk 3]
```diff
@@ -14,20 +14,21 @@
...
 
-class ExecutorAddNotSupportedListener extends H2OSparkListener {
 
-  override def onExecutorAdded(executorAdded: SparkListenerExecutorAdded): Unit = {
-    Log.err(s"Executor ${executorAdded.executorId} without H2O instance discovered, killing the cloud!")
-    H2O.shutdown(-1)
-  }
+class H2OAwareRDD[U: ClassTag](nodes: Seq[String], prev: RDD[U]) extends RDD[U](prev: RDD[U]) {
+
+  override def getPreferredLocations(split: Partition): Seq[String] = nodes
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/rdd/h2o/H2OAwareRDD.scala]: Does this method ensure that the logic will be always executed on the list of specified nodes? Aren't there any edge cases?

Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/rdd/h2o/H2OAwareRDD.scala]: Tests went fine, but still need to check the actual implementation, good questions

File: core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalBackendUtils.scala [hunk 4]
```diff
@@ -134,7 +134,6 @@ private[internal] trait InternalBackendUtils extends SharedBackendUtils {
     // Create global accumulator for list of nodes IP:PORT
     val bc = sc.collectionAccumulator[NodeDesc]
     val isLocal = sc.isLocal
-    val userSpecifiedCloudSize = sc.getConf.getOption("spark.executor.instances").map(_.toInt)
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalBackendUtils.scala]: totally unused variable

File: core/src/main/scala/org/apache/spark/rdd/h2o/H2OAwareRDD.scala (hunk 3 above)
Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/rdd/h2o/H2OAwareRDD.scala]: More context -> It's not guaranteed for 100%. Spark will try to use the preferred locations, but only for duration of `spark.locality.wait` timeout. If the node preferred note is not available till then, it will try to launch the task on different executor.

We can set `spark.locality.wait=0` to ensure we will just try to use the right nodes. This does not affect as as we truly need to run tasks on nodes with H2O( and don't support dynamic allocation) not the other ones - so the behavior will still be the same as it is now

I think we will get rid of this anyway after the big stabilization effort.
